### ISSUE TITLE
[Invoke-ClearAzureSpecificObjects] Something went wrong while clearing the Azure specific objects in the database.

### ISSUE THREAD

**Issue by kai74** [d365collaborative/d365fo.tools#358]
Hi 
I am trying to create a bacpac using "New-D365Backpac

New-D365Bacpac -ExportModeTier2-xxxx -SqlUser xxxxx -BacpacFile D:\DB_backup\update2devdb.bacpac -DatabaseName xxxxxx_axdb_2b18fc3xxxxxx -DatabaseServer xxxxxx.database.windows.net -EnableException -NewDatabaseName update2devdb -ShowOriginalProgress -Verbose 

I am ending up with this message all the time:
_[13:36:39][Invoke-ClearAzureSpecificObjects] Something went wrong while clearing the Azure specific objects in the database. | Exception calling "ExecuteNonQuery" with "0" argument(s): "The database principal owns a schema in the database, and cannot be dropped."
WARNING: [13:36:39][Invoke-ClearAzureSpecificObjects] Stopping because of errors. | Something went wrong while clearing the Azure specific objects in the database.

We are testing with version D365fo.tools : 0.5.64
D365FO is Version A10.0.8 and P Update32 (7.0.5493.16714)

Best regards Kai

**Comment by Splaxi** [comment]
Could you try the
-ExportOnly parameter? 

https://github.com/d365collaborative/d365fo.tools/blob/development/docs/New-D365Bacpac.md#-exportonly

Do you know if your environment is a Self-service or the normal kind where you can RDP into your Tier2 AOS? 

And may I ask why you are using the tools and not LCS to create a bacpac file from your Tier2? 😁

**Comment by kai74** [comment]
We can't use the LCS backup, the size of the database is too big.  600 gb
This is the normal kind, where you can RDP into it. 
I have tested now on an older version  (Version: 10.0.3 (10.0.107.20019) P. Update27 (7.0.5286.41363) and it works perfect (i am not done importing the database, but the export worked perfectly. (Used 12+ hours) 
I have tested with ExportOnly, but it didn't work, it must be something with the version.

In the future hopefully, the lcs function will work.  But when I need to update multiple Dev, test environments this functionality this will really be a great help.

**Comment by Splaxi** [comment]
If you want - I'm happy to help you over a team session, so we can share the screen.

But I'm guessing that there is a new schema in the database, which we need to implement support for.

We are using this script to clear up the database:
https://github.com/d365collaborative/d365fo.tools/blob/development/d365fo.tools/internal/sql/clear-azurebacpacdatabase.sql

And at line 56-71 we handle schemas:

https://github.com/d365collaborative/d365fo.tools/blob/29f341e64b1b39c8077934f2ea74bf7e0d79a1d0/d365fo.tools/internal/sql/clear-azurebacpacdatabase.sql#L56-L71

Could you query the database and see what schemas that are in it?

```
SELECT *
FROM SYS.SCHEMAS
```

**Comment by kai74** [comment]
Hi Motz

 

Thanks for your reply

 

This is the result from the query

 

dbo11
guest22
INFORMATION_SCHEMA33
sys44
ReportingIntegration51
ax61
crt71
ext81
BatchScheduling98
db_owner1638416384
db_accessadmin1638516385
db_securityadmin1638616386
db_ddladmin1638716387
db_backupoperator1638916389
db_datareader1639016390
db_datawriter1639116391
db_denydatareader1639216392
db_denydatawriter1639316393

 

 

You are more than welcome to contact me on Teams. 

You can find me on this email address kai@profiloptik.dk

 

Best regard Kai

**Comment by Splaxi** [comment]
Hi again,

Could you please try and run this:
```
SELECT  s.Name, u.*
FROM    sys.schemas s
        INNER JOIN sys.sysusers u
            ON u.uid = s.principal_id
```

This should show us who owns the schema and make it easier for us to find which one we need to add to the clear script.

**Comment by kai74** [comment]
Hi :)

Name uid status name2 sid roles createdate updatedate altuid password gid
environ hasdbaccess islogin isntname isntgroup isntuser issqluser isaliased
issqlrole isapprole
dbo 1 0 dbo
0x01060000000000640000000000000000EEF3C95D270B7C44A6F971A522949460
NULL 2003-04-08
09:10:42.287 2019-12-07 15:51:07.497 NULL NULL 0 NULL 1 1 0 0 0 1 0 0 0
guest 2 0 guest 0x00 NULL 2003-04-08 09:10:42.317 2003-04-08 09:10:42.317
NULL NULL 0 NULL 0 1 0 0 0 1 0 0 0
INFORMATION_SCHEMA 3 0 INFORMATION_SCHEMA NULL NULL 2009-04-13
12:59:11.717 2009-04-13
12:59:11.717 NULL NULL 0 NULL 0 1 0 0 0 1 0 0 0
sys 4 0 sys NULL NULL 2009-04-13 12:59:11.717 2009-04-13 12:59:11.717 NULL
NULL 0 NULL 0 1 0 0 0 1 0 0 0
ReportingIntegration 1 0 dbo
0x01060000000000640000000000000000EEF3C95D270B7C44A6F971A522949460
NULL 2003-04-08
09:10:42.287 2019-12-07 15:51:07.497 NULL NULL 0 NULL 1 1 0 0 0 1 0 0 0
ax 1 0 dbo
0x01060000000000640000000000000000EEF3C95D270B7C44A6F971A522949460
NULL 2003-04-08
09:10:42.287 2019-12-07 15:51:07.497 NULL NULL 0 NULL 1 1 0 0 0 1 0 0 0
crt 1 0 dbo
0x01060000000000640000000000000000EEF3C95D270B7C44A6F971A522949460
NULL 2003-04-08
09:10:42.287 2019-12-07 15:51:07.497 NULL NULL 0 NULL 1 1 0 0 0 1 0 0 0
ext 1 0 dbo
0x01060000000000640000000000000000EEF3C95D270B7C44A6F971A522949460
NULL 2003-04-08
09:10:42.287 2019-12-07 15:51:07.497 NULL NULL 0 NULL 1 1 0 0 0 1 0 0 0
BatchScheduling 8 0 axdbadmin
0x01060000000100640000000000000000F56D45D5A63D084590CA9895463EABF7
NULL 2019-12-07
16:10:29.103 2019-12-07 16:10:29.107 NULL NULL 0 NULL 1 1 0 0 0 1 0 0 0
db_owner 16384 0 db_owner
0x01050000000000090400000000000000000000000000000000400000 NULL 2003-04-08
09:10:42.333 2009-04-13 12:59:14.467 1 NULL 16384 NULL 0 0 0 0 0 0 0 1 0
db_accessadmin 16385 0 db_accessadmin
0x01050000000000090400000000000000000000000000000001400000 NULL 2003-04-08
09:10:42.333 2009-04-13 12:59:14.467 1 NULL 16385 NULL 0 0 0 0 0 0 0 1 0
db_securityadmin 16386 0 db_securityadmin
0x01050000000000090400000000000000000000000000000002400000 NULL 2003-04-08
09:10:42.350 2009-04-13 12:59:14.467 1 NULL 16386 NULL 0 0 0 0 0 0 0 1 0
db_ddladmin 16387 0 db_ddladmin
0x01050000000000090400000000000000000000000000000003400000 NULL 2003-04-08
09:10:42.350 2009-04-13 12:59:14.467 1 NULL 16387 NULL 0 0 0 0 0 0 0 1 0
db_backupoperator 16389 0 db_backupoperator
0x01050000000000090400000000000000000000000000000005400000 NULL 2003-04-08
09:10:42.350 2009-04-13 12:59:14.467 1 NULL 16389 NULL 0 0 0 0 0 0 0 1 0
db_datareader 16390 0 db_datareader
0x01050000000000090400000000000000000000000000000006400000 NULL 2003-04-08
09:10:42.363 2009-04-13 12:59:14.467 1 NULL 16390 NULL 0 0 0 0 0 0 0 1 0
db_datawriter 16391 0 db_datawriter
0x01050000000000090400000000000000000000000000000007400000 NULL 2003-04-08
09:10:42.363 2009-04-13 12:59:14.467 1 NULL 16391 NULL 0 0 0 0 0 0 0 1 0
db_denydatareader 16392 0 db_denydatareader
0x01050000000000090400000000000000000000000000000008400000 NULL 2003-04-08
09:10:42.380 2009-04-13 12:59:14.467 1 NULL 16392 NULL 0 0 0 0 0 0 0 1 0
db_denydatawriter 16393 0 db_denydatawriter
0x01050000000000090400000000000000000000000000000009400000 NULL 2003-04-08
09:10:42.380 2009-04-13 12:59:14.467 1 NULL 16393 NULL 0 0 0 0 0 0 0 1 0

Den ons. 8. jan. 2020 kl. 08.36 skrev Mötz Jensen <notifications@github.com

**Comment by Splaxi** [comment]
Okay - to get you on with your work, please try this:

You need to locate the path where the d365fo.tools module is installed on the machine.
```
explorer.exe (Split-Path $(Get-Module d365fo.tools -ListAvailable | Select-Object -First 1).Path -Parent)
```` 
![image](https://user-images.githubusercontent.com/8750181/71970089-43645f00-3208-11ea-996a-b8faea2f7fb8.png)

From here you need to locate the `clear-azurebacpacdatabase.sql` file.

It is located in the sub folder: `internal\sql`

![image](https://user-images.githubusercontent.com/8750181/71970028-1fa11900-3208-11ea-9b40-48d6f0bd6b1a.png)

**Important:** You need to start notepad with "Run As Administrator" and then open the `clear-azurebacpacdatabase.sql` file from the location you just found.

On line 62, you need to add the "BatchScheduling" to the list of schemas.
So from this:
```
WHERE SYS.SCHEMAS.NAME IN ('BACKUP','SHADOW')
```

https://github.com/d365collaborative/d365fo.tools/blob/29f341e64b1b39c8077934f2ea74bf7e0d79a1d0/d365fo.tools/internal/sql/clear-azurebacpacdatabase.sql#L62

to this:
```
WHERE SYS.SCHEMAS.NAME IN ('BACKUP','SHADOW','BatchScheduling')
```

Save the `clear-azurebacpacdatabase.sql` and exit all your powershell consoles / sessions.

And now retry the export once more.

**Comment by Splaxi** [comment]
For reference, the default path to the mentioned file is:
C:\Program Files\WindowsPowerShell\Modules\d365fo.tools\0.5.64\internal\sql\clear-azurebacpacdatabase.sql

That should work if you installed the tools directly from PowerShell using the `Install-Module d365fo.tools`

**Comment by kai74** [comment]
Great. I will test once more. And get back to you ASAP.

Den ons. 8. jan. 2020 kl. 11.21 skrev Mötz Jensen <notifications@github.com

**Comment by Splaxi** [comment]
Just wanted to check on you and see if you faced new issues?

**Comment by kai74** [comment]
Hi

It is running the Export now. But not done.. It really takes a loot of
time. But it Is past the old error. I will get back to you when it is done

Thanks a lot for all your help

Kai

**Comment by kai74** [comment]
It used one day and 59 min to export. But the export is done. Will try to import it now.

**Comment by Splaxi** [comment]
Fingers crossed 🤞

**Comment by Splaxi** [comment]
Just wanted to check on you and see if you were able to import the bacpac file?

But I'm going to merge the change into the module this week, based on that you were not able to export the bacpac before and now you are able to export it.

**Comment by kai74** [comment]
Hi Mötz 

 

It is all Done. The environment is up and “spinning” . It used 38 hours to import the database. 

 

So this issue is solved. 

 

I really appreciate all the help.

 

Best regards Kai

**Comment by Splaxi** [comment]
I'm happy that you were able to get it into your environment.

I don't know if you already know, but we actually have a cmdlet / function in place, that can clear table data in a bacpac file. Lets say that you don't want all batch history and stuff like that. Then the tools can remove that from the bacpac file.

It will extract the bacpac file, which is just a zip file and remove the data the desired table(s) and create a new bacpac file. So you will need your 3 times the storage of the size of the bacpac file to be able to handle this. The upside is that you don't need a SQL Server to import it into, clean it up and then export it. So it should be faster than that.

For each GB of data you delete, you will speed up the import process of bacpac file.

**Comment by Splaxi** [comment]
https://github.com/d365collaborative/d365fo.tools/blob/development/docs/Clear-D365TableDataFromBacpac.md

**Comment by kai74** [comment]
Yes, i have seen it.

 

It would be nice if you could make this a parameter. So it can do the cleanup before it starts the export from SQL. Then it will be faster :). It really take some type to unzip the bacpac when it is at this big size :)

 

We only have 30 days old data in the batch history table, but timed saves is time earned 

 

The most significant tables in our database is retail transactions.

 

tableused_mballocated_mb
dbo.RETAILTRANSACTIONSALESTRANS69728.2270182.29
dbo.SALESLINE42045.5242047.38
dbo.LEDGERJOURNALTRANS37386.1637389.61
dbo.TAXTRANS28638.4328639.38
dbo.CUSTTRANS25698.2025700.50
dbo.GENERALJOURNALACCOUNTENTRY25579.5925581.19
dbo.RETAILTRANSACTIONTABLE18579.9618696.79
dbo.INVENTTRANS18393.8818394.91
dbo.INVENTSUMLOGTTS16348.5216349.59
dbo.SOURCEDOCUMENTLINE12172.3612173.17

 

 

And when we look at number of rows

TableNameindexNameRowsTotalPages
INVENTSUMLOGTTSI_7153TTSIDIDX99892173598922
GENERALJOURNALACCOUNTENTRYI_15242RECID63109328525778
INVENTTRANSPOSTINGI_15502DATEVOUCHERTRANSIDX57393634330153
DMFSTAGINGEXECUTIONERRORSI_9142RECID47104982161185
SOURCEDOCUMENTLINEI_7752RECID43219078187242
TAXTRANSGENERALJOURNALACCOUNTENTRYI_1401RECID41736268206194
TAXTRANSI_2292RECID40391317385994
DIMENSIONFOCUSBALANCEI_7445BALANCEQUERYIDX30229919172953
INVENTTRANSI_9569TRANSORIGINIDX23504322243049
INVENTTRANSORIGINI_8551RECID23500786116154
INVENTREPORTDIMHISTORYI_5902TYPETRANSIDREFIDX2201323799666
SALESPARMLINEI_15005RECID21815890235137
RETAILTRANSACTIONSALESTRANSI_13550TRANSACTIONLINEIDX21094472455082

 

 

Best regards Kai

**Comment by Splaxi** [comment]
Without knowing your environment - you should look into the `-CustomSqlFile` parameter of `New-D365Bacpac`

https://github.com/d365collaborative/d365fo.tools/blob/development/docs/New-D365Bacpac.md#-customsqlfile

This allows you to execute a single SQL file, containing whatever fits your needs. This will be executed BEFORE your export.

I'm guessing that you are running on the first versions of a Tier2 environment, where you can logon to the AOS instances using RDP. So this should work for your scenario.

When you are being servered a Self-Service environment, you shouldn't expect this module to be able help you out. Then you would be forced to use the tools available from LCS, but hopefully at that time, they are able to export the bacpac file for. You will end up with what you have now, a bacpac file containing all data. Then you will have to uptake the mentioned "Clear-D365TableDataFromBacpac" cmdlet / function if you want to remove data. That could be because of size or sensitivity.

**Comment by kai74** [comment]
Yes, i have seen it.

 

I will look into it next time :) 

 

Thanks once more.

**Comment by Splaxi** [comment]
This is solved in 0.5.64
